File: fabric/python/fabric/pool.py
import random
import threading


class RelayPool:
    def __init__(self):
        self._lock = threading.Lock()
        self._entries: list[dict] = []  # [{"url": str, "failures": int}]

    def is_empty(self) -> bool:
        with self._lock:
            return len(self._entries) == 0

    def urls(self) -> list[str]:
        with self._lock:
            return [e["url"] for e in self._entries]

    def shuffled_urls(self, n: int) -> list[str]:
        with self._lock:
            random.shuffle(self._entries)
            self._entries.sort(key=lambda e: e["failures"])
            limit = n if n > 0 else len(self._entries)
            if limit > len(self._entries):
                limit = len(self._entries)
            return [self._entries[i]["url"] for i in range(limit)]

    def mark_failed(self, url: str):
        with self._lock:
            for e in self._entries:
                if e["url"] == url:
                    e["failures"] += 1
                    return

    def mark_alive(self, url: str):
        with self._lock:
            for e in self._entries:
                if e["url"] == url:
                    e["failures"] = 0
                    return

    def refresh(self, urls: list[str]):
        with self._lock:
            existing = {e["url"] for e in self._entries}
            for url in urls:
                if url not in existing:
                    self._entries.append({"url": url, "failures": 0})

File: fabric/js/fabric-core/src/pool.ts
interface RelayEntry {
  url: string;
  failures: number;
}

/** Tracks relay URLs with failure counts and provides shuffled selection. */
export class RelayPool {
  private entries: RelayEntry[] = [];

  /** Shuffle entries, sort failures to back, return up to n URLs. */
  shuffledUrls(n: number = Infinity): string[] {
    // Fisher-Yates shuffle
    for (let i = this.entries.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [this.entries[i], this.entries[j]] = [this.entries[j], this.entries[i]];
    }
    // Stable sort by failures ascending
    this.entries.sort((a, b) => a.failures - b.failures);
    return this.entries.slice(0, n).map((e) => e.url);
  }

  markFailed(url: string): void {
    const entry = this.entries.find((e) => e.url === url);
    if (entry) entry.failures++;
  }

  markAlive(url: string): void {
    const entry = this.entries.find((e) => e.url === url);
    if (entry) entry.failures = 0;
  }

  /** Add new URLs to the pool (deduplicates). */
  refresh(urls: Iterable<string>): void {
    const existing = new Set(this.entries.map((e) => e.url));
    for (const url of urls) {
      if (!existing.has(url)) {
        this.entries.push({ url, failures: 0 });
        existing.add(url);
      }
    }
  }

  get isEmpty(): boolean {
    return this.entries.length === 0;
  }

  get urls(): string[] {
    return this.entries.map((e) => e.url);
  }
}
